src/client/easy.c

```diff
@@ -0,0 +1,780 @@
+			   size_t buf_len,
+			   struct node_store_cache *cache)
+{
+	const char *p;
+	size_t n;
+	size_t k;
+	size_t i;
+	const char *addr;
+	const char *id_str;
+	unsigned long long id;
+	const char *role_str;
+	int role;
+	struct client_node_info info;
+
+	p = buf;
+	n = buf_len;
+	while (n > 0) {
+		addr = p;
+		k = strnlen(p, n);
+		if (k == n) {
+			fatal("dqlite_server: parseNodeStore");
+		}
+		p += k + 1;
+		n -= k + 1;
+
+		id_str = p;
+		k = strnlen(p, n);
+		if (k == n) {
+			fatal("dqlite_server: parseNodeStore");
+		}
+		p += k + 1;
+		n -= k + 1;
+		/* Be stricter than strtoull: digits only! */
+		for (i = 0; i < k; i += 1) {
+			if (id_str[i] < '0' || id_str[i] > '9') {
+				fatal("dqlite_server: parseNodeStore");
+			}
+		}
+		errno = 0;
+		id = strtoull(id_str, NULL, 10);
+		if (errno != 0) {
+			fatal("dqlite_server: parseNodeStore");
+		}
+
+		role_str = p;
+		k = strnlen(p, n);
+		if (k == n) {
+			fatal("dqlite_server: parseNodeStore");
+		}
+		p += k + 1;
+		n -= k + 1;
+		if (strcmp(role_str, "spare") == 0) {
+			role = DQLITE_SPARE;
+		} else if (strcmp(role_str, "standby") == 0) {
+			role = DQLITE_STANDBY;
+		} else if (strcmp(role_str, "voter") == 0) {
+			role = DQLITE_VOTER;
+		} else {
+			fatal("dqlite_server: parseNodeStore");
+		}
+
+		info.addr = strdupChecked(addr);
+		info.id = (uint64_t)id;
+		info.role = role;
+		pushNodeInfo(cache, info);
+	}
+}
+
+/* Write the in-memory node store to disk. This discards errors, because:
+ *
+ * - we can't do much to handle any of these error cases
+ * - we don't want to stop everything when this encounters an error, since the
+ *   persisted node store is an optimization, so it's not disastrous for it to
+ *   be missing or out of date
+ * - there is already a "retry" mechanism in the form of the refreshTask thread,
+ *   which periodically tries to write the node store file
+ */
+static void writeNodeStore(struct dqlite_server *server)
+{
+	int store_fd;
+	FILE *f;
+	unsigned i;
+	ssize_t k;
+	const char *role_name;
+	int rv;
+
+	store_fd = openat(server->dir_fd, "node-store-stale",
+			  O_RDWR | O_CREAT | O_TRUNC, 0644);
+	if (store_fd < 0) {
+		return;
+	}
+	f = fdopen(store_fd, "w+");
+	if (f == NULL) {
+		close(store_fd);
+		return;
+	}
+
+	for (i = 0; i < server->cache.len; i += 1) {
+		role_name =
+		    (server->cache.nodes[i].role == DQLITE_SPARE)
+			? "spare"
+			: ((server->cache.nodes[i].role == DQLITE_STANDBY)
+			       ? "standby"
+			       : "voter");
+		k = fprintf(f, "%s%c%" PRIu64 "%c%s%c",
+			    server->cache.nodes[i].addr, '\0',
+			    server->cache.nodes[i].id, '\0', role_name, '\0');
+		if (k < 0) {
+			fclose(f);
+			return;
+		}
+	}
+
+	fclose(f);
+	rv = renameat(server->dir_fd, "node-store-stale", server->dir_fd,
+		      "node-store-fresh");
+	(void)rv;
+}
+
+/**
+ * The format for the local info file consists of concatenated null-terminated
+ * strings: address, then ID (decimal). Example:
+ *
+ * @a\017\0
+ */
```

Comment:
Same as for the node store file, I'd suggest to make it fully textual and add a format version.